Src/WindowsServer/WindowsServer.Shared/Implementation/AssemblyLoader.cs

```diff
@@ -0,0 +1,90 @@
+﻿namespace Microsoft.ApplicationInsights.WindowsServer.Implementation
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Globalization;
+    using System.Reflection;
+
+    /// <summary>
+    /// AssemblyLoader contains logic to load a given assembly and read properties using reflection.
+    /// Inherits MarshalByRefObject so that methods of this class can be executed remotely in separate AppDomain.
+    /// </summary>    
+    internal class AssemblyLoader : MarshalByRefObject
+    {
+        internal string AssemblyName;
+        internal string Culture;
+        internal string PublicKeyToken;
+        internal string[] VersionsToAttempt;
+
+        public AssemblyLoader()
+        {
+            this.AssemblyName = "Microsoft.WindowsAzure.ServiceRuntime";
+            this.Culture = "neutral";
+            this.PublicKeyToken = "31bf3856ad364e35";
+            this.VersionsToAttempt = new string[] { "2.7.0.0", "2.6.0.0"};
+        }
+
+        public bool ReadAndPopulateContextInformation(ref string roleName, ref string roleInstanceId)
+        {
+            Assembly loadedAssembly = null;
+            try
+            {
+                // As this is executed inside a separate AppDomain, it is safe to load assemblies here without interfering with user code.                
+                loadedAssembly = this.AttemptToLoadAssembly(this.AssemblyName, this.Culture, this.PublicKeyToken, this.VersionsToAttempt);
+                if (loadedAssembly != null)
+                {
+                    ServiceRuntime serviceRuntime = new ServiceRuntime();
+                    RoleEnvironment roleEnvironment = serviceRuntime.GetRoleEnvironment(loadedAssembly);
```

Comment:
should you pass assembly class to `ServiceRuntime` constructor, not to `GetRoleEnvironment` logically?